### PR TITLE
Gan tutorial

### PR DESCRIPTION
**Pull request by wangyang59**
I have finished the first draft of gan tutorial

### CODE REVIEW

**Comment by luotao1** [comment]
https://github.com/wangyang59/Paddle/blob/0186edec501a131d3a95c889c38cb0bf480177f1/doc/tutorials/gan/index_en.md 点击文档，会发现前面两幅图太大了，能不能缩小一下。然后三幅图都居中放置吧

File: doc/tutorials/gan/index_en.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Generative Adversarial Networks (GAN) 
+
+This demo implements GAN training described in the original GAN paper (https://arxiv.org/abs/1406.2661) and DCGAN (https://arxiv.org/abs/1511.06434).
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: url信息，可以用 [GAN paper](https://arxiv.org/abs/1406.2661) 显示，不要显示地址。下同。

File: doc/tutorials/gan/index_en.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+The high-level structure of GAN is shown in Figure. 1 below. It is composed of two major parts: a generator and a discriminator, both of which are based on neural networks. The generator takes in some kind of noise with a known distribution and transforms it into an image. The discriminator takes in an image and determines whether it is artificially generated by the generator or a real image. So the generator and the discriminator are in a competitive game in which generator is trying to generate image to look as real as possible to fool the discriminator, while the discriminator is trying to distinguish between real and fake images. 
+
+<center>![](./gan.png)</center>
+<center>Figure 1. GAN-Model-Structure Source: ishmaelbelghazi.github.io/ALI/</center>
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: Figure 标题中不要显示url信息，或者用上面的方式。

File: doc/tutorials/gan/index_en.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+Following the DC-Gan paper (https://arxiv.org/abs/1511.06434), we use convolution/convolution-transpose layer in the discriminator/generator network to better deal with images. The details of the network structures are defined in gan_conf_image.py. 
+
+### Training the model
+To train the GAN model on mnist data, one can use the following command
+```bash
+$python gan_trainer.py -d mnist --useGpu 1
+```
+The generated sample images can be found at ./mnist_samples/ and one example is shown below as Figure 3. 
+<center>![](./mnist_sample.png)</center>
+<center>Figure 2. MNIST Sample</center>
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: 这里是Figure3，序号错了

File: doc/tutorials/gan/index_en.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+## Implementation of GAN Model Structure
+Since GAN model involves multiple neural networks, it requires to use paddle python API. So the code walk-through below can also partially serve as an introduction to the usage of Paddle Python API.
+
+There are three networks defined in gan_conf.py, namely **generator_training**, **discriminator_training** and **generator**. The relationship to the model structure we defined above is that **discriminator_training** is the discriminator, **generator** is the generator, and the **generator_training** combined the generator and discriminator since training generator would require the discriminator to provide loss function. This relationship is described in the following code
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: 最后一个单词code后面加冒号或者句号。下面多处需要修改。

File: doc/tutorials/gan/index_en.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+gen_conf.model_config)
+generator_machine = api.GradientMachine.createFromConfigProto(
+generator_conf.model_config)
+
+# Create trainer
+dis_trainer = api.Trainer.create(dis_conf, dis_training_machine)
+gen_trainer = api.Trainer.create(gen_conf, gen_training_machine)
+```
+
+In order to balance the strength between generator and discriminator, we schedule to train whichever one is performing worse by comparing their loss function value. The loss function value can be calculated by a forward pass through the GradientMachine
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: 句尾加标点。

File: doc/tutorials/gan/index_en.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+copy_shared_parameters(gen_training_machine,
+dis_training_machine)
+copy_shared_parameters(gen_training_machine, generator_machine)
+```
+
+
+## A Toy Example 
+With the infrastructure explained above, we can now walk you through a toy example of generating two dimensional uniform distribution using 10 dimensional Gaussian noise. 
+
+The Gaussian noises are generated using the code below
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: 句尾加标点

File: doc/tutorials/gan/index_en.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+```bash
+$cd data/
+$./get_mnist_data.sh
+```
+
+### Model description
+Following the DC-Gan paper (https://arxiv.org/abs/1511.06434), we use convolution/convolution-transpose layer in the discriminator/generator network to better deal with images. The details of the network structures are defined in gan_conf_image.py. 
+
+### Training the model
+To train the GAN model on mnist data, one can use the following command
```
Comment (luotao1) [inline doc/tutorials/gan/index_en.md]: 句尾加标点

**Comment by wangyang59** [comment]
@luotao1 Are there anything else you would like to me to change? Thanks

**Comment by luotao1** [comment]
@wangyang59 LGTM. How about @Zrachel ?

**Comment by wangyang59** [comment]
@luotao1 Could you please approve and merge this pull request? Thanks a lot~
